### DISCUSSION TITLE
Is Libsla Thread-Safe?

### DISCUSSION THREAD

**Discussion by pjsoberoi** [NationalSecurityAgency/ghidra#6240] [Q&A]
Is libsla thread-safe? Each thread will have it own ContextInternal, Loader, Sleigh translator, DocumentStorage, etc. Should be safe right? Thanks in advance.

**Comment by pjsoberoi** [comment]
I have determined libsla is not is not thread safe. Specifically the following is causing crashes: 

```
        AttributeId::initialize();
        ElementId::initialize();
        sleighroot = docstorage.openDocument(test_params.sla_filename)->getRoot();
```

What do we have to do to make this thread safe? AttributeId and ElementId appear to be working on static global members AFAIK. Wrapping this section of code in a mutex I no longer seem to crash but I kill my performance.

**Comment by astrelsky** [comment]
AFAIK none of the c++ Ghidra code is thread safe.

**Comment by pjsoberoi** [comment]
Using an additional DocumentStorage variable I seem to not be crashing. Need to spend additional time to determine how safe this is.
